bio/picard/markduplicates/wrapper.py

```diff
@@ -5,40 +5,60 @@
 
 
 import tempfile
+from pathlib import Path
 from snakemake.shell import shell
 from snakemake_wrapper_utils.java import get_java_opts
+from snakemake_wrapper_utils.samtools import get_samtools_opts, infer_out_format
 
-log = snakemake.log_fmt_shell()
 
+log = snakemake.log_fmt_shell()
 extra = snakemake.params.get("extra", "")
 # the --SORTING_COLLECTION_SIZE_RATIO default of 0.25 might
 # indicate a JVM memory overhead of that proportion
 java_opts = get_java_opts(snakemake, java_mem_overhead_factor=0.3)
+samtools_opts = get_samtools_opts(snakemake)
+
+
+tool = "MarkDuplicates"
+if snakemake.params.get("withmatecigar", False):
+    tool = "MarkDuplicatesWithMateCigar"
+
 
 bams = snakemake.input.bams
 if isinstance(bams, str):
     bams = [bams]
 bams = list(map("--INPUT {}".format, bams))
 
-if snakemake.output.bam.endswith(".cram"):
+
+output = snakemake.output.bam
+output_fmt = infer_out_format(output)
+convert = ""
+if output_fmt == "CRAM":
     output = "/dev/stdout"
-    if snakemake.params.embed_ref:
-        view_options = "-O cram,embed_ref"
-    else:
-        view_options = "-O cram"
-    convert = f" | samtools view -@ {snakemake.threads} {view_options} --reference {snakemake.input.ref} -o {snakemake.output.bam}"
-else:
-    output = snakemake.output.bam
-    convert = ""
+
+    # NOTE: output format inference should be done by snakemake-wrapper-utils. Keeping it here for backwards compatibility.
+    if snakemake.params.get("embed_ref", False):
+        samtools_opts += ",embed_ref"
+
+    convert = f" | samtools view {samtools_opts}"
+elif output_fmt == "BAM" and snakemake.output.get("idx"):
+    extra += " --CREATE_INDEX"
+
 
 with tempfile.TemporaryDirectory() as tmpdir:
     shell(
-        "(picard MarkDuplicates"  # Tool and its subcommand
+        "(picard {tool}"  # Tool and its subcommand
         " {java_opts}"  # Automatic java option
         " {extra}"  # User defined parmeters
         " {bams}"  # Input bam(s)
         " --TMP_DIR {tmpdir}"
         " --OUTPUT {output}"  # Output bam
         " --METRICS_FILE {snakemake.output.metrics}"  # Output metrics
-        " {convert} ) {log}"  # Logging
+        " {convert}) {log}"  # Logging
     )
+
+
+output_prefix = Path(snakemake.output.bam).with_suffix("")
+if snakemake.output.get("idx"):
+    if output_fmt == "BAM" and snakemake.output.idx != str(output_prefix) + ".bai":
+        shell("mv {output_prefix}.bai {snakemake.output.idx}")
```

Comment:
I don't think the second check is necessary. If the `mv` doesn't do anything, that's no problem.
```suggestion
if snakemake.output.get("idx") and output_fmt == "BAM":
        shell("mv {output_prefix}.bai {snakemake.output.idx}")
```

Reply:
I don` think you can move a file to itself. Destination needs to be different, no?

Reply:
You are right. :see_no_evil: Feel free to either move all conditions into one row (with another `and`) or just merge as is. Whichever you like better.